android/app/src/main/kotlin/coop/polypoly/polypod/updatenotification/UpdateNotification.kt

```diff
@@ -5,23 +5,65 @@ import coop.polypoly.polypod.Preferences
 import coop.polypoly.polypod.R
 
 class UpdateNotification(private val context: Context) {
-    val id = context.resources.getInteger(R.integer.update_notification_id)
+    companion object {
+        data class MockData(var id: Int? = null)
+        val mockData = MockData()
+    }
+
+    private enum class State {
+        NOT_SEEN,
+        PUSH_SEEN,
+        ALL_SEEN;
+
+        companion object {
+            fun parse(s: String?) =
+                s.let { name -> values().firstOrNull { it.name == name } }
+        }
+    }
+
+    val id = mockData.id
+        ?: context.resources.getInteger(R.integer.update_notification_id)
+
     val title = context.getString(R.string.update_notification_title)
+
     val text = context.getString(R.string.update_notification_text)
+
     val pushDelay =
         context.resources.getInteger(R.integer.update_notification_push_delay)
 
-    val inAppNotificationSeen: Boolean
-        get() = id == 0 || Preferences.getSeenInAppNotificationId(context) == id
+    val showPush: Boolean
+        get() = state == State.NOT_SEEN
+
+    val showInApp: Boolean
+        get() = state != State.ALL_SEEN
 
-    val pushNotificationSeen: Boolean
-        get() = id == 0 || Preferences.getSeenPushNotificationId(context) == id
+    private var state: State = loadLastState()
+        set(value) {
+            field = value
+            Preferences.setLastNotification(context, id, state.name)
+        }
+
+    private fun loadLastState(): State {
+        val (lastId, lastState) = Preferences.getLastNotification(context)
+        return when (id) {
+            0 -> State.ALL_SEEN
+            lastId -> State.parse(lastState) ?: State.ALL_SEEN
+            else -> State.NOT_SEEN
+        }
+    }
 
-    fun markInAppNotificationSeen() {
-        Preferences.setSeenInAppNotificationId(context, id)
+    fun onFirstRun() {
+        state = State.ALL_SEEN
     }
 
-    fun markPushNotificationSeen() {
-        Preferences.setSeenPushNotificationId(context, id)
+    fun onPushSeen() {
+        if (state == State.NOT_SEEN)
+            state = State.PUSH_SEEN
```

Comment:
Else? would staying at ALL_SEEN or PUSH_SEEN be valid?

Reply:
Yeah, if it's already in `PUSH_SEEN`, we don't need to set it to `PUSH_SEEN`. If it's already in `ALL_SEEN`, we wouldn't want the in app notification to show up _again_, even though it's unexpected for `onPushSeen` to be invoked in that state.

Reply:
So maybe an exception in that case?

Reply:
You mean when `onPushSeen` is invoked when we expected it wouldn't? Doesn't seem like that'd be something `UpdateNotification` would worry about - its job is to react to events and maintain consistency of the stored data...